crates/nu-cli/src/env/environment.rs

```diff
@@ -51,9 +54,38 @@ impl Environment {
         Environment {
             environment_vars: env,
             path_vars: path,
+            direnv: DirectorySpecificEnvironment::new(configuration.nu_env_dirs()),
         }
     }
 
+    pub fn maintain_directory_environment(&mut self) -> std::io::Result<()> {
+        self.direnv.env_vars_to_delete()?.iter().for_each(|k| {
+            self.remove_env(&k);
+        });
+        self.direnv.env_vars_to_add()?.iter().for_each(|(k, v)| {
+            self.add_env(&k, &v, true);
+        });
+
+        self.direnv
+            .overwritten_values_to_restore()?
+            .iter()
+            .for_each(|(k, v)| {
+                self.add_env(&k, &v, true);
+            });
```

Comment:
I wasn't sure - if we change from one directory that overwrites "MYENVVAR" to another one that overwrites it to a different value, if we restore here do we end up overwriting the new value?

Reply:
I tested one scenario for this.
The directory `test` contains a `.nu` with the var `overwritten = test`
The directory `test2`, a subfolder of `test`, contains a `.nu` with `overwritten = test2`
The directory `test3` is a subfolder of `test2` and does not contain anything.

So `test` writes a variable, `test2` overwrites it and then we can study the result in `test3`.

```
/home/sam/github/test> echo $nu.env | get overwritten
test
/home/sam/github/test> cd test2
/home/sam/github/test/test2> echo $nu.env | get overwritten
test2
/home/sam/github/test/test2> cd test3
/home/sam/github/test/test2/test3> echo $nu.env | get overwritten
test2
```

It works like this because we keep track of which directory overwrote what variables, so we know that if we are no longer in a directory, we should restore those values. Since we are still in `test2`, we know that we should not restore the values overwritten by `test2`.

This was the only situation I could think of where this overwriting and restoring could cause issues, but there could very well be more.